### PR TITLE
feat(translator): Security policies for TLS gateways

### LINKED ISSUE #6704: Security Policies for TLS Gateways
*Description*:

SecurityPolicies don't currently support TLSRoute's. I'd like to propose allowing SecurityPolicies for TLSRoutes. Specifically client IP based auth from Security Policies. 

*Related*:
* https://github.com/envoyproxy/gateway/issues/4908
* https://github.com/envoyproxy/gateway/pull/6496

Example Policy:

```yaml
apiVersion: gateway.envoyproxy.io/v1alpha1
kind: SecurityPolicy
metadata:
  name: restrict-client-ips-tlsroute-example
  namespace: envoy-gateway-system
spec:
  targetRefs:
    - group: gateway.networking.k8s.io
      kind: TLSRoute
      name: tlsroute-example
  authorization:
    defaultAction: Deny
    rules:
    - action: Allow
      principal:
        clientCIDRs:
        - <public_address>
```

Known Issues:

A known bug with [the RBAC network filter](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#role-based-access-control-rbac-proto) https://github.com/envoyproxy/envoy/issues/9023 allows connections to the upstream to be initiated even if the rbac policy evaluates to `DENY`. This can be mitigated by instead using `filter_chain_match` with the source_prefix_range filter. However, doing that I believe might not allow the user to have a SecurityPolicy `defaultAction` other than `Deny`. 

Workarounds:

I can and have tested this with `EnvoyPatchPolicy`, however, I would love official support in `SecurityPolicy` or another similar CR if possible. 

I am willing to work on adding this if the team is interested in this.

### PR DESCRIPTION
**Pull request by asHasnain**
**What this PR does / why we need it**:
Adds SecurityPolicy support for TLSRoute with client IP-based authorization. Currently, SecurityPolicies support Gateway, HTTPRoute, GRPCRoute, and TCPRoute, but not TLSRoute. This PR extends authorization support to TLSRoute (passthrough mode), allowing users to apply client IP allow/deny lists using SecurityPolicy.

**Which issue(s) this PR fixes**:

Fixes https://github.com/envoyproxy/gateway/issues/6704

---

**PR Checklist**

- [X] **Authorship & ownership**: Coding agents / AI assistants are welcome, but I have reviewed every change, understand how and why it works, can explain and maintain it, and take full responsibility for this PR. I have not submitted generated output I do not understand.
- [X] **DCO**: All commits are signed off (`git commit -s`). See [DCO: Sign your work](https://gateway.envoyproxy.io/community/contributing/#dco-sign-your-work).
- [ ] **API agreed first**: If this PR contains API changes (changes under `/api`), the API was discussed and agreed **before** the implementation. The API change can be in a separate PR, or in the same PR, but the API must be agreed before implementation. N/A if this PR does not contain API changes.
- [X] **Required checks pass**: `make generate gen-check`, `make lint`, and the unit-test/coverage build pass. (Flaky e2e failures are not considered breakages, but `gen-check`, `lint`, and coverage **MUST** pass.)
- [ ] **Tests added/updated**: New/changed code is covered by appropriate tests. N/A if this PR does not contain code changes.
- [ ] **Docs**: User-facing changes update the [docs](https://github.com/envoyproxy/gateway/tree/main/site), either in this PR or a follow-up PR. N/A if this PR does not contain user-facing changes.
- [ ] **Release notes**: For any non-trivial change, added a release-note fragment under `release-notes/current/<section>/<pr-number>-<slug>.md` (see `release-notes/current/README.md` for sections and naming). N/A if this PR does not contain non-trivial changes.
- [X] **Generated files committed**: Ran `make gen-check` and committed the result if API/helm charts/modules changed.
- [ ] **Scope & compatibility**: The PR is reasonably scoped (no unrelated changes) and preserves backward compatibility, or any breaking change is called out above and documented in `release-notes/current/breaking_changes/`.
- [ ] **Codex review**: Requested a Codex review and addressed all of its comments.
- [ ] **Copilot review**: Requested a Copilot review and addressed all of its comments.

### CODE REVIEW

File: api/v1alpha1/securitypolicy_types.go [hunk 1]
```diff
@@ -37,19 +37,19 @@ type SecurityPolicy struct {
 
 // SecurityPolicySpec defines the desired state of SecurityPolicy.
 //
-// NOTE: SecurityPolicy can target Gateway, ListenerSet, HTTPRoute, GRPCRoute, and TCPRoute.
-// When a SecurityPolicy targets a TCPRoute, only client-IP CIDR based authorization
+// NOTE: SecurityPolicy can target Gateway, ListenerSet, HTTPRoute, GRPCRoute, and TCPRoute, and TLSRoute.
```
Comment (guydc) [inline api/v1alpha1/securitypolicy_types.go:40]: Can you please add a release note as well for this new feautre?

File: test/e2e/tests/tlsroute_authorization_client_ip.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+			testTLSRouteConnectionBlocked(t, gwAddr, "blocked.example.com")
+		})
+
+		t.Run("allowed client IP can connect to allowed.example.com", func(t *testing.T) {
+			testTLSRouteConnectionAllowed(t, gwAddr, "allowed.example.com")
+		})
+	},
+}
+
+func testTLSRouteConnectionBlocked(t *testing.T, gwAddr, hostname string) {
```
Comment (guydc) [inline test/e2e/tests/tlsroute_authorization_client_ip.go:78]: Please enhance the test to also confirm that the RBAC filter actually blocked/allowed the traffic, using the RBAC stats and prometheus test utils: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/rbac_filter#statistics

**Review by guydc — COMMENTED** [review]
Thanks @asHasnain for picking this up, it's a useful enhancement of security policy for additional TCP/TLS Scenerios. 

Can we add some yaml tests in the GW-API layer covering translation of more complex scenarios like multiple rules in a single TLS route and multiple TLS routes in a listener? 
TCP routes usually have a simpler xds shape than TLS routes, so I just want to make sure that reusing the existing translation of security policy for TCP covers all cases.

File: test/e2e/tests/tlsroute_authorization_client_ip.go (hunk 2 above)
Comment (asHasnain) [reply test/e2e/tests/tlsroute_authorization_client_ip.go:78]: A function is added to verify it via RBAC metrics `envoy_rbac_denied` and `envoy_rbac_allowed`.

File: api/v1alpha1/securitypolicy_types.go (hunk 1 above)
Comment (asHasnain) [reply api/v1alpha1/securitypolicy_types.go:40]: A release note is added.

**Review by asHasnain — COMMENTED** [review]
Thanks @guydc for reviewing. I've added tests for the specified scenarios.

**Comment by guydc** [comment]
cc @zhaohuabing

**Comment by zhaohuabing** [comment]
Hi @asHasnain TestE2E/TLSRouteAuthzWithClientIP/allowed_client_IP_can_connect_to_allowed.example.co is failing. Could you please fix it?

**Comment by asHasnain** [comment]
> Hi @asHasnain TestE2E/TLSRouteAuthzWithClientIP/allowed_client_IP_can_connect_to_allowed.example.co is failing. Could you please fix it?

Sorry, I was on vacation so could not get back earlier.

The test passed locally for me:
```
=== RUN   TestE2E/TLSRouteAuthzWithClientIP/blocked_client_IP_cannot_connect_to_blocked.example.com
    tlsroute_authorization_client_ip.go:87: Connection blocked as expected: EOF
    tlsroute_authorization_client_ip.go:191: RBAC filter denied 1 connections (confirmed via Prometheus)
=== RUN   TestE2E/TLSRouteAuthzWithClientIP/allowed_client_IP_can_connect_to_allowed.example.com
    tlsroute_authorization_client_ip.go:160: Successfully received response: HTTP/1.1 200 OK
        Content-Type: application/json
        X-Content-Type-Options: nosniff
        Date: Thu, 16 Jul 2026 18:24:12 GMT
        Content-Length: 439
        
        {
         "path": "/",
         "host": "allowed.example.com",
         "method": "GET",
         "proto": "HTTP/1.1",
         "headers": {
          "Accept": [
           "*/*"
          ],
          "User-Agent": [
           "test-client"
          ]
         },
         "httpPort": "3000",
         "namespace": "gateway-conformance-infra",
         "ingress": "",
         "service": "tls-backend",
         "pod": "tls-backend-5d64b6c68b-5zxdh",
         "tls": {
          "version": "TLSv1.3",
          "serverName": "allowed.example.com",
          "cipherSuite": "TLS_AES_128_GCM_SHA256"
         }
        }
    tlsroute_authorization_client_ip.go:189: RBAC filter allowed 1 connections (confirmed via Prometheus)
=== NAME  TestE2E/TLSRouteAuthzWithClientIP
    timing.go:87: 2026-07-16T20:24:12.573236+02:00: Test TLSRouteAuthzWithClientIP completed in 72ms
```
However, the pipeline failure might be related to the short RBAC metrics scraping timeout (10s) in the test. I've increased timeout to 1m (like [another test](https://github.com/asHasnain/gateway/blob/main/test/e2e/tests/circuitbreaker.go#L93)). For verification, I checked the envoy proxy pod stats and it showed that the RBAC metric values were correctly updated:
```
tls-passthrough-8443.rbac.allowed: 4
tls-passthrough-8443.rbac.denied: 4
tls-passthrough-8443.rbac.shadow_allowed: 0
tls-passthrough-8443.rbac.shadow_denied: 0
```
The stats in Prometheus format (`/stats/prometheus`) did not show `namespace` so I've removed it from RBAC metrics query.
[stats_2026-08-05.txt](https://github.com/user-attachments/files/30775307/eg_pr_9464_passing_e2e_test_2026-08-05_stats.txt)
[e2e_test_output_2026-08-05.txt](https://github.com/user-attachments/files/30775309/eg_pr_9464_passing_e2e_test_2026-08-05.txt)
[e2e_test_output_2026-07-16.txt](https://github.com/user-attachments/files/30775305/eg_pr_9464_passing_e2e_test_2026-07-16.txt)

File: internal/gatewayapi/securitypolicy.go [hunk 3]
```diff
@@ -991,16 +991,26 @@ func (t *Translator) translateSecurityPolicyForRoute(
 
 		irKey := t.getIRKey(gtwCtx.Gateway)
 		switch route.GetRouteType() {
-		case resource.KindTCPRoute:
+		case resource.KindTCPRoute, resource.KindTLSRoute:
```
[the remark this answers is not in the document]
Comment (asHasnain) [reply internal/gatewayapi/securitypolicy.go:1384]: I'll address it by blocking sectionName on TCPRoute/TLSRoute targets at admission via CEL validation rather than fixing the metadata propagation. Since neither `TCPRoute` nor `TLSRoute` has per-rule enforcement boundary at the XDS filter-chain level (all rules pool into one ir.TCPRoute/one filter chain), `sectionName` targeting is semantically invalid for both kinds. The CEL rules reject the combination with a clear error, so the silent-failure path in translateSecurityPolicyForRoute (target.SectionName != nil && ...) is now unreachable for these route types.

[the remark this answers is not in the document]
Comment (guydc) [reply internal/gatewayapi/securitypolicy.go:1384]: @codex - WDYT? Is that sufficient to cover this edge case?

**Review by zhaohuabing — COMMENTED** [review]
Hi @asHasnain, this looks good. 

I just have one question: would this also work for `Terminate` mode `TLSRoute`?

If yes, can we add tests for `Termination` mode and update the PR description?

**Comment by asHasnain** [comment]
@zhaohuabing Yes, it works for Terminate mode too — both modes produce the same ir.TCPRoute and go through the same SecurityPolicy translation path. The current tests only cover Passthrough so I will add a unit test fixture for Terminate mode (and update the PR description accordingly).

**Comment by zhaohuabing** [comment]
Hi @asHasnain is it possible to also cover terminate mode in the e2e test?

**Comment by asHasnain** [comment]
> Hi @asHasnain is it possible to also cover terminate mode in the e2e test?

Yes, it's doable — in this PR or a follow-up? A couple of things I'd confirm before writing it:
The Terminate listener needs a cert/key in a kubernetes.io/tls Secret. Is it fine to add one inline in gateway-conformance-infra (same namespace as the Gateway, no ReferenceGrant — matches tlsroute-tls-termination.yaml) or do you prefer cross-namespace + ReferenceGrant?
And after terminating at the listener, forward plain TCP to infra-backend-v1:8080 (the existing echo server)?

**Comment by zhaohuabing** [comment]
> > Hi @asHasnain is it possible to also cover terminate mode in the e2e test?
> 
> Yes, it's doable — in this PR or a follow-up? A couple of things I'd confirm before writing it: The Terminate listener needs a cert/key in a kubernetes.io/tls Secret. Is it fine to add one inline in gateway-conformance-infra (same namespace as the Gateway, no ReferenceGrant — matches tlsroute-tls-termination.yaml) or do you prefer cross-namespace + ReferenceGrant? And after terminating at the listener, forward plain TCP to infra-backend-v1:8080 (the existing echo server)?

I'd to include the test in this PR to verify the SP for teminate mode.
A inline Secret and echo server sound good to me.

Thanks!
